File: combinatorics/resource/LDL/C++/bai4_AdjacencyMatrix.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cout << "Nhap so dinh n: ";
    cin >> n;

    // Nhap ma tran ke
    cout << "Nhap ma tran ke:\n";
    vector<vector<int>> matrix(n, vector<int>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> matrix[i][j];
        }
    }

    // Adjacency List: luu danh sach ke
    vector<vector<int>> adj_list(n);

    // Edge List: luu danh sach canh (u, v, k)
    struct Edge {
        int u, v, k;
    };
    vector<Edge> edge_list;

    // Adjacency Map: map moi dinh v -> (incoming, outgoing)
    // incoming[v][u] = vector cac canh tu u toi v
    // outgoing[v][w] = vector cac canh tu v toi w
    vector<unordered_map<int, vector<pair<int, int>>>> incoming(n), outgoing(n);

    // Duyet qua ma tran ke
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            int k = matrix[i][j]; // so canh tu i den j
            if (k > 0) {
                // Them vao adjacency list (lap k lan neu co nhieu canh)
                for (int t = 0; t < k; t++) {
                    adj_list[i].push_back(j);
                }

                // Them vao edge list
                edge_list.push_back({i, j, k});

                // Cap nhat adjacency map (incoming va outgoing)
                for (int t = 0; t < k; t++) {
                    outgoing[i][j].push_back({i, j}); // canh tu i toi j
                    incoming[j][i].push_back({i, j}); // canh toi j tu i
                }
            }
        }
    }

    // In danh sach ke
    cout << "\n============ Adjacency List:\n";
    for (int u = 0; u < n; u++) {
        cout << u << ": ";
        for (auto v : adj_list[u]) cout << v << " ";
        cout << "\n";
    }

    // In danh sach canh
    cout << "\n============ Edge List:\n";
    for (auto &e : edge_list) {
        cout << e.u << " " << e.v << " " << e.k << "\n";
    }

    // In adjacency map
    cout << "\n============ Adjacency Map:\n";
    for (int v = 0; v < n; v++) {
        cout << "Dinh " << v << ":\n";

        // Incoming
        cout << "  Incoming:\n";
        for (auto &kv : incoming[v]) {
            cout << "    " << kv.first << ": ";
            for (auto &edge : kv.second) {
                cout << "(" << edge.first << "," << edge.second << ") ";
            }
            cout << "\n";
        }

        // Outgoing
        cout << "  Outgoing:\n";
        for (auto &kv : outgoing[v]) {
            cout << "    " << kv.first << ": ";
            for (auto &edge : kv.second) {
                cout << "(" << edge.first << "," << edge.second << ") ";
            }
            cout << "\n";
        }
    }

    return 0;
}

File: combinatorics/resource/LDL/Python/bai4_AdjacencyMatrix.py
from collections import defaultdict

# Nhap ma tran ke (do thi co huong)
n = int(input("Nhap so dinh n: "))
print("Nhap ma tran ke:")
matrix = []
for _ in range(n):
    row = list(map(int, input().split()))
    matrix.append(row)

# Adjacency List
adj_list = defaultdict(list)

# Edge List
edge_list = []

# Adjacency Map: moi dinh se map toi (incoming, outgoing)
adj_map = defaultdict(lambda: (defaultdict(list), defaultdict(list)))

for i in range(n):
    for j in range(n):
        k = matrix[i][j]  # So canh tu i den j
        if k > 0:
            adj_list[i].extend([j] * k)        # Them j vao danh sach ke cua i, lap k lan
            edge_list.append((i, j, k))        # Them (i, j, k): tu i den j co k canh

            for _ in range(k):
                adj_map[i][1][j].append((i, j))  # Outgoing tu i den j
                adj_map[j][0][i].append((i, j))  # Incoming den j tu i

# In danh sach ke
print("\n============Adjacency List:")
for u in range(n):
    print(f"{u}: {adj_list[u]}")

# In danh sach canh
print("\n============Edge List:")
for u, v, k in edge_list:
    print(f"{u} {v} {k}")

# In Adjacency Map
print("\n============Adjacency Map:")
for v in range(n):
    incoming, outgoing = adj_map[v][0], adj_map[v][1]
    print(f"Dinh {v}:")
    print("  Incoming:")
    for u in incoming:
        print(f"    {u}: {incoming[u]}")
    print("  Outgoing:")
    for w in outgoing:
        print(f"    {w}: {outgoing[w]}")